### PR TITLE
Add experimental signInWithIdToken for Apple, Google

### PR DESCRIPTION
**Pull request by DanMossa**
Recreates  https://github.com/supabase/gotrue-js/pull/603

This is the endpoint we need to hit for native auth. i'm also going to be working on native auth in the supabase-flutter repo.

Here's a WIP of the PR

https://github.com/supabase/supabase-flutter/blob/signInWithIdToken/README.md#native-auth-example

### CODE REVIEW

**Comment by DanMossa** [comment]
I'm not sure why I keep getting
`TypeError: Cannot reconstruct a Request with a used body` When testing this out. Any ideas?
@dshukertjr @bdlukaa

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+## [1.4.3]
+
+- feat: add support for `signInWithIdToken`.
+-
```
Comment (dshukertjr) [inline CHANGELOG.md:4]:
```suggestion

```

File: lib/src/gotrue_client.dart [hunk 2]
```diff
@@ -216,6 +216,52 @@ class GoTrueClient {
         redirectTo: redirectTo, scopes: scopes, queryParams: queryParams);
   }
 
+  /// Allows signing in with an ID token issued by certain supported providers.
+  /// The [idToken] is verified for validity and a new session is established.
+  ///
+  /// This method is experimental.
+  Future<AuthResponse> signInWithIdToken({
```
Comment (dshukertjr) [inline lib/src/gotrue_client.dart:223]: We can probably use the experimental annotation here! https://api.flutter.dev/flutter/meta/experimental-constant.html
```suggestion
  @Experimental()
  Future<AuthResponse> signInWithIdToken({
```

**Review by dshukertjr — COMMENTED** [review]
Thank you so much for working on this! I've updated the version to 1.5.0 since we are adding a feature. Let me know what you think about the experimental flag and if it's good, we can merge it and release it!

File: lib/src/gotrue_client.dart (hunk 2 above)
Comment (Vinzent03) [reply lib/src/gotrue_client.dart:223]: I think you would need to add the `meta` package for this.

Comment (DanMossa) [reply lib/src/gotrue_client.dart]: Done!

File: lib/src/gotrue_client.dart [hunk 3]
```diff
@@ -216,6 +217,51 @@ class GoTrueClient {
...
+  @Experimental()
+  Future<AuthResponse> signInWithIdToken({
+    required Provider provider,
+    required String idToken,
+    String? nonce,
+    String? captchaToken,
+  }) async {
+    _removeSession();
+
+    if (provider != Provider.google && provider != Provider.apple) {
```
Comment (bdlukaa) [inline lib/src/gotrue_client.dart:231]: I believe this must be in documentation, because Github and Facebook sign in is pretty common

Comment (DanMossa) [reply lib/src/gotrue_client.dart:231]: Like in the comments about the method? Inside the `///` ?

**Comment by DanMossa** [comment]
> Thank you so much for working on this! I've updated the version to 1.5.0 since we are adding a feature. Let me know what you think about the experimental flag and if it's good, we can merge it and release it!

Oof. Now getting run time error due to adding the `@Experimental()` tag. I'll look into it in a bit.

**Comment by dshukertjr** [comment]
@DanMossa Sorry, I provided the wrong information. It was `@experimental` instead of `@Experimental()`. Also, currently it seems like the `@experimental` annotation doesn't really do much in terms of the analyzer, so added back the comments you had about the method being experimental as well!

**Comment by DanMossa** [comment]
@dshukertjr @bdlukaa  Done!

**Review by dshukertjr — COMMENTED** [review]
@DanMossa Just a quick check. Why did we remove the `@experimental` flag?

https://api.flutter.dev/flutter/meta/experimental-constant.html

**Comment by DanMossa** [comment]
> @DanMossa Just a quick check. Why did we remove the `@experimental` flag?
> 
> https://api.flutter.dev/flutter/meta/experimental-constant.html

I think I misunderstood what you meant by your comment above then?

https://github.com/supabase/gotrue-dart/pull/119#issuecomment-1418437339

**Comment by dshukertjr** [comment]
@DanMossa Ah, sorry for not being clear! In [this comment](https://github.com/supabase/gotrue-dart/pull/119#discussion_r1096738066), I provided the wrong information suggesting to use `@Experimental()`, which broke all the tests. I then corrected myself [here](https://github.com/supabase/gotrue-dart/pull/119#issuecomment-1418437339) saying we should use `@experimental`, and not `@Experimental()`, and I also pushed my commit to update the code. 

I think we should still have the `@experimental` flag, because even though it's not doing much at the moment, the Dart team might add warnings on the IDE based on it in the future!

**Comment by DanMossa** [comment]
@dshukertjr Ohh that makes perfect sense. If you want, you can make those changes again and merge whenever!

**Comment by dshukertjr** [comment]
@DanMossa Will do that right now and release this version!

**Review by dshukertjr — APPROVED** [review]

**Comment by dshukertjr** [comment]
@DanMossa gotrue-dart v1.5.0 has been published 🎉 Thanks for all the work here!

**Comment by hmarat** [comment]
@dshukertjr @DanMossa Hey! Firstly thank you for working on this important feature! So is signing with idToken now working properly for both Apple & Google sign-in?

**Comment by theniceboy** [comment]
It turns out this feature actually won't work unless some changes made in the main supabase backend:

https://github.com/supabase/supabase-flutter/issues/5#issuecomment-1426634366

https://github.com/supabase/supabase-flutter/issues/5#issuecomment-1430699456

**Comment by DanMossa** [comment]
> It turns out this feature actually won't work unless some changes are made in the main supabase backend:
> 
> https://github.com/supabase/supabase-flutter/issues/5#issuecomment-1426634366
> 
> https://github.com/supabase/supabase-flutter/issues/5#issuecomment-1430699456

This is not true.

This feature does work without implementing backend features. I'm using it now. I just have it so that iOS users only see sign in with apple and Android users only see sign in with Google.

Both see email.

**Comment by theniceboy** [comment]
> This feature does work without implementing backend features. I'm using it now. I just have it so that iOS users only see sign in with apple and Android users only see sign in with Google.

@DanMossa 

I see. You're suggesting that by using native sign-in, **you'd have to give up the sign-in provider on all but one platform**. I would say that's not really acceptable because most of the Flutter apps are targeting multiple platforms.

Currently If you configure any platform to use sign-in with google, you will be using the OAuth client ID for web, because the current Supabase sign-in flow always use a browser. If you want to use this native sign-in feature right now on let's say iOS, that means you have to set the OAuth client ID in your Supabase dashboard to use the **iOS** client ID. If you do that, native sign-in with Google will indeed work on your iOS app, however, none of your other platforms, such as Web, Android, Desktop, can even implement sign-in with Google, because the current Supabase backend only allows one client ID per auth provider.

**Pull request merged**